### PR TITLE
feat: add sql completions powered by LLM

### PR DESCRIPTION
**Pull request by jczhong84**
Adding another AI assistant feature, SQL Complete, to help query authoring like the Github Copilot

Currently we're using this codemirror [extension](https://github.com/asadm/codemirror-copilot), we can reimplement it to meet our additional requrements in the future if needed. 

The prompt and context still needs to be tuned for better quality.

Snip20241114_2

By default it is disabled, user can enable it in the user settings.
image

### CODE REVIEW

File: querybook/webapp/hooks/queryEditor/extensions/useSqlCompleteExtension.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+import { AICommandType, AISocketEvent } from 'const/aiAssistant';
+import aiAssistantSocket from 'lib/ai-assistant/ai-assistant-socketio';
+
+/**
+ * Remove overlapping part with prefix from completion.
+ *
+ * e.g. removeOverlapPrefix('sel', 'elect *') => 'ect *'
+ */
+const removeOverlapPrefix = (completion: string, prefix: string) => {
```
Comment (zhangvi7) [inline querybook/webapp/hooks/queryEditor/extensions/useSqlCompleteExtension.ts:12]: just curious why is this needed?

**Review by zhangvi7 — COMMENTED** [review]
Looks good! Does the code mirror support showing multi line suggestions back to user?

File: querybook/webapp/hooks/queryEditor/extensions/useSqlCompleteExtension.ts (hunk 1 above)
Comment (jczhong84) [reply querybook/webapp/hooks/queryEditor/extensions/useSqlCompleteExtension.ts:12]: It is a hack workaround. found sometimes it will have overlap with the prefix. I'll add some comments there.

**Comment by jczhong84** [comment]
> Looks good! Does the code mirror support showing multi line suggestions back to user?

yes, it supports multiple lines. 
image

File: querybook/webapp/hooks/queryEditor/extensions/useSqlCompleteExtension.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    let commonPrefixLength = 0;
+    for (let i = 0; i < prefix.length; i++) {
+        if (completion.startsWith(prefix.slice(i))) {
+            commonPrefixLength = prefix.length - i;
+            break;
+        }
+    }
+    if (commonPrefixLength > 0 || prefix === '') {
```
Comment (czgu) [inline querybook/webapp/hooks/queryEditor/extensions/useSqlCompleteExtension.ts:20]: if prefix == 0 then commonPrefixLength = 0 so the slice would be a noop

**Review by czgu — APPROVED** [review]
